File: python/33. Search in Rotated Sorted Array.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l, r = 0, len(nums) - 1
        while l + 1 < r:
            mid = (l + r) / 2
            if nums[mid] == target:
                return mid
            else:
                if nums[mid] > nums[l]:
                    if target >= nums[l] and target < nums[mid]:
                        r = mid
                    else:
                        l = mid
                else:
                    if target >= nums[mid] and target <= nums[r]:
                        l = mid
                    else:
                        r = mid
        if nums[l] == target:
            return l
        if nums[r] == target:
            return r
        return -1

File: cpp/1-50/33. Search in Rotated Sorted Array.cpp
class Solution {
public:
	int search(vector<int>& nums, int target) {
		int low = 0;
		int high = nums.size() - 1;
		while(low <= high)
		{
			int mid = (low + high) / 2;
			int num = (target < nums[0]) == (nums[mid] < nums[0])?
						nums[mid]: target < nums[0]? INT_MIN:INT_MAX;
			if(num < target)
				low = mid + 1;
			else if(num > target)
				high = mid - 1;
			else
			    return mid;
		}
		return -1;
	}
};